### ISSUE TITLE
Complete support for second saturation derivatives

### ISSUE THREAD

**Issue by Astrinus** [CoolProp/CoolProp#581]
I'd be grateful to CoolProp developers if they will support derivatives like \d^2(y)/\dP\dH or \d^2(y)/\dH^2.

**Comment by ibell** [comment]
To be frank I am not sure how to do those derivatives ALONG the saturation curve.  What do you need this for?

**Comment by Astrinus** [comment]
A linearization for a ORC plant model.

**Comment by ibell** [comment]
@thorade Since you are the expert in this field, any idea how to calculate these other derivatives ALONG the saturation curves?

**Comment by thorade** [comment]
Two more questions:
* Are you talking about derivatives ALONG the saturation line or derivatives in the two phase region? 
* Are you interested in pure fluids or in mixtures?

**Comment by ibell** [comment]
The derivatives are ALONG the saturation line, not in the two-phase region.

For now lets say pure fluids, once you get into mixture derivatives
everything becomes very messy indeed.

**Comment by thorade** [comment]
I was asking because the examples in the original question are derivatives wrt enthalpy. I am not sure if you can have derivatives wrt enthalpy ALONG the saturation line. Second derivatives along the sat line wrt temperature/pressure for pure fluids should be possible.

**Comment by ibell** [comment]
Yes, precisely my problem.  In fact this comes into the splines in the two
phase region that we implemented as well, and I am fairly sure that what I
did is incorrect.  Ok enough, but not correct.  I feel like you might be
able to derive the derivatives, but I hit a wall.

I've done the second derivatives with temperature or pressure.

What I think you could do is to do  dy/dT|sigma / dz/dT|sigma to get
dy/dz|sigma, right?  Perhaps you could then expand those?  A first thought.

**Comment by thorade** [comment]
I spent today with learning some Python+CoolProp, the next days I'll be away from my computer, but once I return, I will continue to work on this. After thinking about it, I am convinced that also the other derivatives should be possible to calculate analytically

**Comment by ibell** [comment]
Welcome to the python club!  Please let me know if you make progress on
this issue.

**Comment by thorade** [comment]
I think I do now know how to calculate all derivatives along the sat line.
A couple of examples are here:
https://github.com/CoolProp/CoolProp/blob/master/doc/notebooks/Saturation.ipynb
http://nbviewer.ipython.org/github/CoolProp/CoolProp/blob/master/doc/notebooks/Saturation.ipynb

I could write up the equations in LaTeX, but that would take some more days. Alternatively I could scan & send you my handwritten notes

**Comment by ibell** [comment]
Do you want to write this up in a more formal way?  We could work on a paper maybe even?  At the least, it would be nice to have the LaTeX.  I'm struggling a bit to follow all the equations.  You know you can put equations into markdown in cells in the notebook, right?

I've pushed this back for version 5.2, which will have a good number of new things, including, hopefuly, this.

**Comment by thorade** [comment]
OK, I'll write down the relevant equations in LaTeX or in the notebook.

**Comment by thorade** [comment]
@Astrinus does that cover everything you are interested in?
Do you need mixed derivatives?

**Comment by Astrinus** [comment]
I think that's all... maybe. I have implemented central differences to bypass the problem.
I am sorry I do not remember what I needed because several months passed from the last time I worked on the model.

**Comment by ibell** [comment]
@thorade Nice work with the notebook.  That looks like it will cover derivatives w.r.t T or p quite nicely up to second order.  I'm still at a bit of a loss as to how we would do mixed derivatives like d^2(T)/(dP dH) along saturation, but I guess that is enough of an edge case to be not too important.

**Comment by thorade** [comment]
I added another example, d^2(s)/(dh dP), just the equation, not a verification plot.

**Comment by thorade** [comment]
Related: https://github.com/jupyter/nbviewer/issues/547

**Comment by ibell** [comment]
Closing as part of an issue-tracker cleanup — this issue has had no activity in over two years.

*To be transparent: this comment and the close action are being performed by Claude (Anthropic's AI coding assistant), running on behalf of @ibell. Ian reviewed and approved the bulk-close criteria; I am the one executing the action.*

If you would like this issue reconsidered, please feel free to **re-open it** — a fresh comment with current behaviour against the latest CoolProp release is the most useful way to revive the discussion.

— Claude (acting for @ibell)
